kong/runloop/plugins_iterator.lua

```diff
@@ -123,86 +162,89 @@ local function get_next(self)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if route_id and consumer_id and combos[COMBO_RC] then
-        plugin_configuration = load_configuration(self, name, route_id, nil, consumer_id)
-        if plugin_configuration then
-          break
-        end
+    if route_id and consumer_id and combos[COMBO_RC] then
+      plugin_configuration = load_configuration(ctx, name, route_id, nil, consumer_id)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if service_id and consumer_id and combos[COMBO_SC] then
-        plugin_configuration = load_configuration(self, name, nil, service_id, consumer_id)
-        if plugin_configuration then
-          break
-        end
+    if service_id and consumer_id and combos[COMBO_SC] then
+      plugin_configuration = load_configuration(ctx, name, nil, service_id, consumer_id)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if route_id and service_id and combos[COMBO_RS] then
-        plugin_configuration = load_configuration(self, name, route_id, service_id, nil)
-        if plugin_configuration then
-          break
-        end
+    if route_id and service_id and combos[COMBO_RS] then
+      plugin_configuration = load_configuration(ctx, name, route_id, service_id, nil)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if consumer_id and combos[COMBO_C] then
-        plugin_configuration = load_configuration(self, name, nil, nil, consumer_id)
-        if plugin_configuration then
-          break
-        end
+    if consumer_id and combos[COMBO_C] then
+      plugin_configuration = load_configuration(ctx, name, nil, nil, consumer_id)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if route_id and combos[COMBO_R] then
-        plugin_configuration = load_configuration(self, name, route_id, nil, nil)
-        if plugin_configuration then
-          break
-        end
+    if route_id and combos[COMBO_R] then
+      plugin_configuration = load_configuration(ctx, name, route_id, nil, nil)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if service_id and combos[COMBO_S] then
-        plugin_configuration = load_configuration(self, name, nil, service_id, nil)
-        if plugin_configuration then
-          break
-        end
+    if service_id and combos[COMBO_S] then
+      plugin_configuration = load_configuration(ctx, name, nil, service_id, nil)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if combos[COMBO_GLOBAL] then
-        plugin_configuration = load_configuration(self, name, nil, nil, nil)
-      end
+    if combos[COMBO_GLOBAL] then
+      plugin_configuration = load_configuration(ctx, name, nil, nil, nil)
+    end
 
-    until true
+    ::after_combos::
```

Comment:
Can this be moved inside that next `if` clause? Perhaps also rename it to `::configuration_found::`.

Reply:
It can also mean `nothing found` - if `combos` is nil, you go here too. If we name it something like `found` it would be misleading.

Reply:
in not combos, then we should perhaps jump over the whole block?

Reply:
e.g. replace this `do` with:
https://github.com/Kong/kong/blob/10d2e9c70e1f48565e520175be24c5a634143ff1/kong/runloop/plugins_iterator.lua#L152

```
if not combos then
```

Reply:
> Then all the gotos will only jump when config was found.

Not really, you can go through every `if` block there and still not have a plugin configuration found at the point of the `found` label (that's why there's an `if` right after that).

Reply:
As @bungle [originally suggested](https://github.com/Kong/kong/pull/4594#discussion_r283080554), I think changing the whole `goto` dance with 

```
local plugin_configuration
if combos then
  plugin_configuration = try_combos(ctx, combos, name, route_id, service_id, consumer_id)
end
```

should simplify things a lot. @kikito, I understand your intuition about the cost of a function call, but LuaJIT is a trace compiler and the one thing trace compilers are really good at is getting rid of function calls. And given @lukego's recent comments about the weirdness of `goto` behavior, I really don't think we should assume that making a plain call there would be necessarily slower.

Reply:
@hishamhm it is changed now according to my comments, and `if` can't jump inside `if`. so that check is still there, it was never meant to be removed, see my original comment, but you cannot but label inside if.

Reply:
@bungle oh, what I meant was just that now, as of 6c17b13, it is possible to reach label `found` with the config not really being found at that point, so the label is misleading.

Reply:
Ok, changing to function calls